modules/messages/src/lib.rs

```diff
@@ -641,7 +662,37 @@ decl_module! {
 			};
 			if let Some(confirmed_messages) = confirmed_messages {
 				// handle messages delivery confirmation
-				T::OnDeliveryConfirmed::on_messages_delivered(&lane_id, &confirmed_messages);
+				let preliminary_callback_overhead = relayers_state.total_messages.saturating_mul(
+					single_message_callback_overhead
+				);
+				let actual_callback_weight = T::OnDeliveryConfirmed::on_messages_delivered(
+					&lane_id,
+					&confirmed_messages,
+				);
+				match preliminary_callback_overhead.checked_sub(actual_callback_weight) {
+					Some(difference) if difference == 0 => (),
+					Some(difference) => {
+						log::trace!(
+							target: "runtime::bridge-messages",
+							"Messages delivery callback has returned unspent weight to refund the submitter: \
+							{} - {} = {}",
+							preliminary_callback_overhead,
+							actual_callback_weight,
+							difference,
+						);
+						actual_weight = actual_weight - difference;
+					},
+					None => {
+						debug_assert!(false, "The delivery confirmation callback is wrong");
```

Comment:
This is something I'm not comfortable with, but Substrate pallets are full of debug_asserts, so it should be fine

Reply:
Wow, interesting. If that's the convention, let's keep it - I guess it's there just to know that it is never triggered in tests (Assuming they are run in debug mode not release)?